binaries/autotuner.cc

```diff
@@ -0,0 +1,550 @@
+        caffe2::FLAGS_input_dims.size() || caffe2::FLAGS_input_type.size()) {
+      CAFFE_ENFORCE_NE(
+          0,
+          caffe2::FLAGS_input_dims.size(),
+          "Input dims must be specified when input files are not specified.");
+      CAFFE_ENFORCE_NE(
+          0,
+          caffe2::FLAGS_input_type.size(),
+          "Input types must be specified when input files are not specified.");
+
+      vector<string> input_dims_list =
+          caffe2::split(';', caffe2::FLAGS_input_dims);
+      CAFFE_ENFORCE_EQ(
+          input_names.size(),
+          input_dims_list.size(),
+          "Input names and input dims should have the same number of items.");
+      vector<string> input_type_list =
+          caffe2::split(';', caffe2::FLAGS_input_type);
+      CAFFE_ENFORCE_EQ(
+          input_names.size(),
+          input_type_list.size(),
+          "Input names and input types should have the same number of items.");
+      for (size_t i = 0; i < input_names.size(); ++i) {
+        vector<string> input_dims_str = caffe2::split(',', input_dims_list[i]);
+        vector<int> input_dims;
+        for (const string& s : input_dims_str) {
+          input_dims.push_back(caffe2::stoi(s));
+        }
+        if (input_type_list[i] == "uint8_t") {
+          inputs.push_back(InputDescription(input_names[i], input_dims, false));
+        } else if (input_type_list[i] == "float") {
+          inputs.push_back(InputDescription(input_names[i], input_dims, true));
+        } else {
+          CAFFE_THROW(
+              "Unsupported input type ",
+              input_type_list[i],
+              " for input ",
+              input_names[i]);
+        }
+      }
+    } else {
+      CAFFE_THROW(
+          "You requested input tensors, but neither input_file nor input_dims + input_type is set.");
+    }
+  }
+
+  /* Load all parameters into a workspace, so we can lookup their shapes when
+   * needed. */
+  unique_ptr<caffe2::Workspace> paramWorkspace(new caffe2::Workspace());
+  caffe2::NetDef init_net;
+  CAFFE_ENFORCE(ReadProtoFromFile(caffe2::FLAGS_init_net, &init_net));
+  CAFFE_ENFORCE(paramWorkspace->RunNetOnce(init_net));
+
+  caffe2::NetDef input_net;
+  CAFFE_ENFORCE(ReadProtoFromFile(caffe2::FLAGS_net, &input_net));
+  caffe2::NetDef best_net(input_net);
+
+  for (int op_index = 0; op_index < input_net.op_size(); op_index++) {
+    const caffe2::OperatorDef& op = input_net.op(op_index);
+
+    if (op.type() == "Conv") {
+      CAFFE_ENFORCE(
+          op.input_size() >= 2, "Conv operator must have 2 or 3 inputs");
+      caffe2::Blob* kernel_blob = paramWorkspace->GetBlob(op.input(1));
+      CAFFE_ENFORCE(
+          kernel_blob != nullptr,
+          "Weights blob ",
+          op.input(1),
+          " for Conv operator ",
+          op.name(),
+          " is not initialized");
+      const caffe2::TensorCPU kernel_tensor =
+          kernel_blob->Get<caffe2::TensorCPU>();
+
+      if (kernel_tensor.ndim() == 4) {
```

Comment:
I'm not entirely sure why you need the logic to check everything before attempting NNPACK -- wouldn't it be simpler to just try every Conv? or are there cases where the backend would actually crash and ruin the autotuning